File: client/app/src/test/java/com/gbndt/shijiaoqi/platform/pouch/PouchActivateTest.kt
package com.gbndt.shijiaoqi.platform.pouch

import com.gbndt.shijiaoqi.platform.crypt.Wm2
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.UUID

class PouchActivateTest {
    @Test
    fun activateThenOpenProcessById() {
        val (p, client) = loggedPouch()
        val body = """{"current":180}""".toByteArray()
        val proc = processMember("工艺", body)
        val snap = projectSnap(client, "工程", proc)
        p.cacheClosure(snap)
        p.activate(snap.assetId)
        assertArrayEquals(body, p.openProcess(proc.id))
        assertReject(Pouch.ERR_NOT_FOUND) { p.openProcess(snap.assetId) }
    }

    @Test
    fun rejectsIncompleteMismatchCopyFullWeldAndOfflineCurrent() {
        val (p, client) = loggedPouch()
        val proc = processMember("工艺", """{"a":1}""".toByteArray())
        val good = projectSnap(client, "工程", proc)
        p.cacheClosure(good)

        val root = good.members.first()
        val missing = good.copy(
            members = listOf(root),
            digest = Digest.closureSum(listOf(Digest.Member(root.id, root.revision, root.digest, root.content))),
        )
        assertReject(Pouch.ERR_INCOMPLETE) { p.cacheClosure(missing) }

        val mismatch = projectSnap(client, "串版", processMember("工艺", """{"a":1}""".toByteArray()))
        val mismatchRoot = mismatch.members.first()
        val brokenRoot = mismatchRoot.copy(deps = listOf(mismatchRoot.deps.first().copy(revision = 99)))
        assertReject(Pouch.ERR_MISMATCH) { p.cacheClosure(mismatch.copy(members = listOf(brokenRoot) + mismatch.members.drop(1))) }

        val tamperedProc = processMember("工艺", """{"a":1}""".toByteArray())
        val tampered = projectSnap(client, "摘要", tamperedProc)
        val dirty = tampered.members.toMutableList()
        dirty[0] = dirty[0].copy(content = "tampered".toByteArray())
        assertReject(Pouch.ERR_INTEGRITY) { p.cacheClosure(tampered.copy(members = dirty)) }

        val copied = projectSnap(UUID.randomUUID(), "只拷", processMember("他机", """{"b":1}""".toByteArray()))
        assertReject(Pouch.ERR_FORBIDDEN) { p.cacheClosure(copied) }

        p.setPolicy(1, Pouch.SCOPE_ALL)
        val second = projectSnap(client, "第二份", processMember("工艺2", """{"c":1}""".toByteArray()))
        assertReject(Pouch.ERR_CACHE_FULL) { p.cacheClosure(second) }

        p.setPolicy(2, Pouch.SCOPE_ALL)
        p.cacheClosure(second)
        p.activate(good.assetId)
        p.setWelding(true)
        assertReject(Pouch.ERR_FORBIDDEN) { p.activate(second.assetId) }
        assertEquals(good.assetId, p.activeProject())
        p.setWelding(false)

        p.setPolicy(2, Pouch.SCOPE_CURRENT)
        p.setOnline(false)
        assertReject(Pouch.ERR_NOT_FOUND) { p.activate(UUID.randomUUID()) }
    }

    @Test
    fun currentScopeDropsOtherProjects() {
        val (p, client) = loggedPouch()
        p.setPolicy(2, Pouch.SCOPE_CURRENT)
        val aProc = processMember("A工艺", """{"a":1}""".toByteArray())
        val bProc = processMember("B工艺", """{"b":1}""".toByteArray())
        val a = projectSnap(client, "A", aProc)
        val b = projectSnap(client, "B", bProc)
        p.cacheClosure(a)
        p.cacheClosure(b)
        p.activate(b.assetId)
        assertFalse(p.hasClosure(a.assetId))
        assertThrows(NoSuchElementException::class.java) { p.open(aProc.id) }
        assertArrayEquals("""{"b":1}""".toByteArray(), p.openProcess(bProc.id))
    }

    @Test
    fun personalActivateFollowsOwner() {
        val unwrap = Wm2.randomKey()
        val who = UUID.randomUUID()
        val client = UUID.randomUUID()
        val p = Pouch()
        p.login(unwrap, who, false)
        p.bindClient(client)
        val proc = processMember("工艺", """{"p":1}""".toByteArray())
        val snap = projectSnap(client, "我的工程", proc)
        val root = snap.members.first().copy(level = Pouch.LEVEL_PERSONAL, ownerId = who)
        p.cacheClosure(snap.copy(level = Pouch.LEVEL_PERSONAL, members = listOf(root) + snap.members.drop(1)))
        p.logout()
        p.login(unwrap, UUID.randomUUID(), false)
        p.bindClient(client)
        assertReject(Pouch.ERR_FORBIDDEN) { p.activate(snap.assetId) }
    }

    @Test
    fun cacheTransitThenActivate() {
        val unwrap = Wm2.randomKey()
        val dek = Wm2.randomKey()
        val who = UUID.randomUUID()
        val factoryId = UUID.randomUUID()
        val client = UUID.randomUUID()
        val procBody = """{"current":180}""".toByteArray()
        val proc = processMember("工艺", procBody)
        val snap = projectSnap(client, "工程", proc)
        val fid = Pouch.uuidBytes(factoryId)
        val cid = Pouch.uuidBytes(client)
        val sealed = snap.members.map { m ->
            TransitMember(
                id = m.id,
                level = m.level,
                name = m.name,
                revision = m.revision,
                ownerId = m.ownerId,
                content = Wm2.seal(dek, m.content, Wm2.clientTransitAad(fid, cid, Pouch.uuidBytes(m.id), m.revision)),
                kind = m.kind,
                status = m.status,
                digest = m.digest,
                deps = m.deps,
            )
        }
        val wrap = Wm2.seal(unwrap, dek, Wm2.clientTransitDekAad(fid, cid))
        val p = Pouch()
        p.login(unwrap, who, false)
        p.bindClient(client)
        p.cacheTransit(
            factoryId,
            client,
            TransitClosure(
                wrap = wrap,
                members = sealed,
                assetId = snap.assetId,
                revision = snap.revision,
                kind = snap.kind,
                level = snap.level,
                status = snap.status,
                digest = snap.digest,
                targetClientId = client,
            ),
        )
        p.activate(snap.assetId)
        assertArrayEquals(procBody, p.openProcess(proc.id))
    }

    private fun assertReject(code: String, block: () -> Unit) {
        val e = assertThrows(PouchRejected::class.java, block)
        assertEquals(code, e.code)
    }

    private fun loggedPouch(): Pair<Pouch, UUID> {
        val p = Pouch()
        p.login(Wm2.randomKey(), UUID.randomUUID(), false)
        val client = UUID.randomUUID()
        p.bindClient(client)
        return p to client
    }

    private fun processMember(name: String, body: ByteArray): ClosureMemberPlain {
        return ClosureMemberPlain(
            id = UUID.randomUUID(),
            kind = Pouch.KIND_PROCESS,
            level = Pouch.LEVEL_FACTORY,
            name = name,
            status = Pouch.STATUS_AVAILABLE,
            revision = 1,
            content = body,
            digest = Digest.sum(body),
        )
    }

    private fun projectSnap(client: UUID, name: String, vararg procs: ClosureMemberPlain): ClosureSnapshotPlain {
        val deps = procs.map { AssetDep(it.id, it.revision, it.digest) }
        val body = """{"items":[]}""".toByteArray()
        val rootId = UUID.randomUUID()
        val root = ClosureMemberPlain(
            id = rootId,
            kind = Pouch.KIND_PROJECT,
            level = Pouch.LEVEL_FACTORY,
            name = name,
            status = Pouch.STATUS_AVAILABLE,
            revision = 1,
            content = body,
            digest = Digest.sum(body),
            deps = deps,
        )
        val members = listOf(root) + procs
        val pack = Digest.closureSum(members.map { Digest.Member(it.id, it.revision, it.digest, it.content) })
        return ClosureSnapshotPlain(
            kind = Pouch.KIND_PROJECT,
            assetId = rootId,
            revision = 1,
            level = Pouch.LEVEL_FACTORY,
            status = Pouch.STATUS_AVAILABLE,
            digest = pack,
            targetClientId = client,
            members = members,
        )
    }
}

File: factory/server/internal/service/pouch_activate_test.go
// C3：按 cache_scope 与上限激活恰好一份；processId 解工艺；缺成员/串版/只拷/超上限/焊接中拒绝。
package service_test

import (
	"bytes"
	"errors"
	"testing"

	"github.com/google/uuid"

	"wmesh/factory/internal/platform/contentcrypt"
	"wmesh/factory/internal/platform/digest"
	"wmesh/factory/internal/platform/domain"
	"wmesh/factory/internal/platform/id"
	factory "wmesh/factory/internal/service"
	"wmesh/factory/internal/store"
)

func TestPouchActivateAndOpenProcess(t *testing.T) {
	p, client, _, _ := loggedPouch(t)
	procBody := []byte(`{"current":180}`)
	proc := processMember("工艺", procBody)
	snap := projectSnap(client, "工程", proc)
	if err := p.CacheClosure(snap); err != nil {
		t.Fatal(err)
	}
	if err := p.Activate(snap.AssetID); err != nil {
		t.Fatal(err)
	}
	got, err := p.OpenProcess(proc.ID)
	if err != nil || !bytes.Equal(got, procBody) {
		t.Fatalf("process %s %v", got, err)
	}
	if _, err := p.OpenProcess(snap.AssetID); !errors.Is(err, domain.ErrNotFound) {
		t.Fatalf("project as process: %v", err)
	}
}

func TestPouchActivateRejectsIncompleteMismatchCopyFullWeldOffline(t *testing.T) {
	p, client, _, _ := loggedPouch(t)
	proc := processMember("工艺", []byte(`{"a":1}`))
	good := projectSnap(client, "工程", proc)
	if err := p.CacheClosure(good); err != nil {
		t.Fatal(err)
	}

	missing := good
	root := missing.Members[0]
	missing.Members = []factory.ClosureMember{root}
	missing.Digest = digest.ClosureSum([]digest.Member{{
		ID: root.ID, Revision: root.Revision, Digest: root.Digest, Content: root.Content,
	}})
	if err := p.CacheClosure(missing); !errors.Is(err, domain.ErrClosureIncomplete) {
		t.Fatalf("missing: %v", err)
	}

	mismatch := projectSnap(client, "串版", processMember("工艺", []byte(`{"a":1}`)))
	mismatch.Members[0].Deps[0].Revision = 99
	if err := p.CacheClosure(mismatch); !errors.Is(err, domain.ErrClosureMismatch) {
		t.Fatalf("mismatch: %v", err)
	}

	tampered := projectSnap(client, "摘要", processMember("工艺", []byte(`{"a":1}`)))
	tampered.Members[0].Content = []byte("tampered")
	if err := p.CacheClosure(tampered); !errors.Is(err, domain.ErrIntegrity) {
		t.Fatalf("integrity: %v", err)
	}

	other := id.New()
	copied := projectSnap(other, "只拷", processMember("他机", []byte(`{"b":1}`)))
	if err := p.CacheClosure(copied); !errors.Is(err, domain.ErrForbidden) {
		t.Fatalf("copy: %v", err)
	}

	second := projectSnap(client, "第二份", processMember("工艺2", []byte(`{"c":1}`)))
	if err := p.CacheClosure(second); err != nil {
		t.Fatal(err)
	}
	if err := p.Activate(good.AssetID); err != nil {
		t.Fatal(err)
	}
	p.SetWelding(true)
	if err := p.Activate(second.AssetID); !errors.Is(err, domain.ErrForbidden) {
		t.Fatalf("weld switch: %v", err)
	}
	if p.ActiveProject() == nil || *p.ActiveProject() != good.AssetID {
		t.Fatal("active moved while welding")
	}
	p.SetWelding(false)

	p.SetPolicy(2, store.CacheScopeCurrent)
	p.SetOnline(false)
	if err := p.Activate(id.New()); !errors.Is(err, domain.ErrNotFound) {
		t.Fatalf("offline current: %v", err)
	}
}

func TestPouchCurrentScopeDropsOtherProjects(t *testing.T) {
	p, client, _, _ := loggedPouch(t)
	p.SetPolicy(2, store.CacheScopeCurrent)
	aProc := processMember("A工艺", []byte(`{"a":1}`))
	bProc := processMember("B工艺", []byte(`{"b":1}`))
	a := projectSnap(client, "A", aProc)
	b := projectSnap(client, "B", bProc)
	if err := p.CacheClosure(a); err != nil {
		t.Fatal(err)
	}
	if err := p.CacheClosure(b); err != nil {
		t.Fatal(err)
	}
	if err := p.Activate(b.AssetID); err != nil {
		t.Fatal(err)
	}
	if p.HasClosure(a.AssetID) {
		t.Fatal("current kept other project")
	}
	if _, err := p.Open(aProc.ID); !errors.Is(err, domain.ErrNotFound) {
		t.Fatalf("dropped member: %v", err)
	}
	got, err := p.OpenProcess(bProc.ID)
	if err != nil || !bytes.Equal(got, []byte(`{"b":1}`)) {
		t.Fatalf("kept %s %v", got, err)
	}
}

func TestPouchActivatePersonalFollowsOwner(t *testing.T) {
	p, client, who, key := loggedPouch(t)
	proc := processMember("工艺", []byte(`{"p":1}`))
	snap := projectSnap(client, "我的工程", proc)
	snap.Level = store.AssetLevelPersonal
	snap.Members[0].Level = store.AssetLevelPersonal
	snap.Members[0].CreatorID = &who
	if err := p.CacheClosure(snap); err != nil {
		t.Fatal(err)
	}
	p.Logout()
	other := id.New()
	if err := p.Login(key, other, false); err != nil {
		t.Fatal(err)
	}
	p.BindClient(client)
	if err := p.Activate(snap.AssetID); !errors.Is(err, domain.ErrNotFound) {
		t.Fatalf("personal: %v", err)
	}
}

func loggedPouch(t *testing.T) (*factory.Pouch, uuid.UUID, uuid.UUID, []byte) {
	t.Helper()
	p := factory.NewPouch()
	key, err := contentcrypt.RandomKey()
	if err != nil {
		t.Fatal(err)
	}
	who := id.New()
	if err := p.Login(key, who, false); err != nil {
		t.Fatal(err)
	}
	client := id.New()
	p.BindClient(client)
	return p, client, who, key
}

func processMember(name string, body []byte) factory.ClosureMember {
	return factory.ClosureMember{
		ID: id.New(), Kind: factory.KindProcess, Level: factory.AssetLevelFactory, Name: name,
		Status: factory.AssetAvailable, Revision: 1, Content: body, Digest: digest.Sum(body),
	}
}

func projectSnap(client uuid.UUID, name string, procs ...factory.ClosureMember) factory.ClosureSnapshot {
	deps := make([]factory.AssetDep, len(procs))
	for i, p := range procs {
		deps[i] = factory.AssetDep{ID: p.ID, Revision: p.Revision, Digest: p.Digest}
	}
	body := []byte(`{"items":[]}`)
	root := factory.ClosureMember{
		ID: id.New(), Kind: factory.KindProject, Level: factory.AssetLevelFactory, Name: name,
		Status: factory.AssetAvailable, Revision: 1, Content: body, Digest: digest.Sum(body), Deps: deps,
	}
	cid := client
	return sealSnap(factory.KindProject, root, procs, nil, &cid)
}
